pkg/logutil/zap_text_encoder.go

```diff
@@ -0,0 +1,608 @@
+		final.EncodeLevel(ent.Level, final)
+		if cur == final.buf.Len() {
+			// User-supplied EncodeLevel was a no-op. Fall back to strings to keep
+			// output JSON valid.
+			final.AppendString(ent.Level.String())
+		}
+		final.endQuoteFiled()
+	}
+
+	if ent.LoggerName != "" && final.NameKey != "" {
+		final.beginQuoteFiled()
+		cur := final.buf.Len()
+		nameEncoder := final.EncodeName
+
+		// if no name encoder provided, fall back to FullNameEncoder for backwards
+		// compatibility
+		if nameEncoder == nil {
+			nameEncoder = zapcore.FullNameEncoder
+		}
+
+		nameEncoder(ent.LoggerName, final)
+		if cur == final.buf.Len() {
+			// User-supplied EncodeName was a no-op. Fall back to strings to
+			// keep output JSON valid.
+			final.AppendString(ent.LoggerName)
+		}
+		final.endQuoteFiled()
+	}
+	if ent.Caller.Defined && final.CallerKey != "" {
+		final.beginQuoteFiled()
+		cur := final.buf.Len()
+		final.EncodeCaller(ent.Caller, final)
+		if cur == final.buf.Len() {
+			// User-supplied EncodeCaller was a no-op. Fall back to strings to
+			// keep output JSON valid.
+			final.AppendString(ent.Caller.String())
+		}
+		final.endQuoteFiled()
+	}
+	// add Message
+	if len(ent.Message) > 0 {
+		final.beginQuoteFiled()
+		final.AppendString(ent.Message)
+		final.endQuoteFiled()
+	}
+	if enc.buf.Len() > 0 {
+		final.addElementSeparator()
+		final.buf.Write(enc.buf.Bytes())
+	}
+	addFields(final, fields)
+	final.closeOpenNamespaces()
+	if ent.Stack != "" && final.StacktraceKey != "" {
+		final.beginQuoteFiled()
+		final.AddString(final.StacktraceKey, ent.Stack)
+		final.endQuoteFiled()
+	}
+
+	if final.LineEnding != "" {
+		final.buf.AppendString(final.LineEnding)
+	} else {
+		final.buf.AppendString(zapcore.DefaultLineEnding)
+	}
+
+	ret := final.buf
+	putTextEncoder(final)
+	return ret, nil
+}
+
+func (enc *textEncoder) truncate() {
+	enc.buf.Reset()
+}
+
+func (enc *textEncoder) closeOpenNamespaces() {
+	for i := 0; i < enc.openNamespaces; i++ {
+		enc.buf.AppendByte('}')
+	}
+}
+
+func (enc *textEncoder) addKey(key string) {
+	enc.addElementSeparator()
+	enc.safeAddStringWithQuote(key)
+	enc.buf.AppendByte('=')
+}
+
+func (enc *textEncoder) addElementSeparator() {
+	last := enc.buf.Len() - 1
+	if last < 0 {
+		return
+	}
+	switch enc.buf.Bytes()[last] {
+	case '{', '[', ':', ',', ' ', '=':
+		return
+	default:
+		enc.buf.AppendByte(',')
+	}
+}
+
+func (enc *textEncoder) appendFloat(val float64, bitSize int) {
+	enc.addElementSeparator()
+	switch {
+	case math.IsNaN(val):
+		enc.buf.AppendString("NaN")
+	case math.IsInf(val, 1):
+		enc.buf.AppendString("+Inf")
+	case math.IsInf(val, -1):
+		enc.buf.AppendString("-Inf")
+	default:
+		enc.buf.AppendFloat(val, bitSize)
+	}
+}
+
+// safeAddString JSON-escapes a string and appends it to the internal buffer.
+// Unlike the standard library's encoder, it doesn't attempt to protect the
+// user from browser vulnerabilities or JSONP-related problems.
+func (enc *textEncoder) safeAddString(s string) {
```

Comment:
How about solutions in https://stackoverflow.com/questions/28595664/how-to-stop-json-marshal-from-escaping-and ? Then you don't need to self-implement a JSON encoder.

Reply:
@breeswish The main reason for not using the `encoding/json` of the standard library is the implementation of the `encoding/json` has reflecting overhead , we already know the type is a string. and this implement is according to the [`json_encoder`](https://github.com/uber-go/zap/blob/master/zapcore/json_encoder.go) in zap, check it with the `encoding/json`,  the JSON-escape implementation is same.